src/pyodide-py/pyodide/_importhooks.py

```diff
@@ -0,0 +1,80 @@
+from ._core import JsProxy
+from importlib.machinery import ExtensionFileLoader
+from importlib.abc import MetaPathFinder
+from importlib.util import spec_from_loader
+import sys
+
+# From Python glossary: An importer is "both a finder and loader object."
+class JsImporter(MetaPathFinder, ExtensionFileLoader):
+    jsproxies = {}
+    ### Finder methods
+    @classmethod
+    def find_spec(cls, fullname, path, target=None):
+        [parent, _, child] = fullname.rpartition(".")
+        if parent:
+            parent_module = sys.modules[parent]
+            if not hasattr(parent_module, "__loader__") or not isinstance(
+                parent_module.__loader__, JsImporter
+            ):
+                # Not one of us.
+                return None
+            try:
+                jsproxy = getattr(parent_module, child)
+            except AttributeError:
+                raise ModuleNotFoundError(
+                    f"No module named {fullname!r}", name=fullname
+                ) from None
+            if not isinstance(jsproxy, JsProxy):
+                raise ModuleNotFoundError(
+                    f"No module named {fullname!r}", name=fullname
+                )
+        else:
+            try:
+                jsproxy = JsImporter.jsproxies[fullname]
+            except KeyError:
+                return None
+        loader = cls(fullname, jsproxy)
+        return spec_from_loader(fullname, loader, origin="javascript")
+
+    def __init__(self, name, jsproxy):
+        super().__init__(name, None)
+        self.jsproxy = jsproxy
+
+    def __repr__(self) -> str:
+        return "javascript module"
+
+    @classmethod
+    def find_module(cls, fullname, path):
+        raise NotImplementedError(
+            "find_module() is deprecated in favor of find_spec() since Python 3.4"
+        )
+
+    @staticmethod
+    def invalidate_caches() -> None:
+        pass
+
```

Comment:
I believe both of these can be left out; they have default implementations: https://github.com/python/cpython/blob/3.8/Lib/importlib/abc.py
```suggestion
```